File: finance_tracker/rules.py
from __future__ import annotations

import re
from dataclasses import asdict, dataclass, field
from datetime import date, datetime
from decimal import Decimal
from typing import Any, Iterable

from .models import Transaction, money


SCHEMA_VERSION = 1
STAGE_ORDER = {
    "TRANSACTION_NORMALIZATION": 10,
    "VENDOR_NORMALIZATION": 20,
    "CLASSIFICATION": 30,
    "TAGGING": 40,
    "EVIDENCE": 50,
    "CASHBACK": 60,
}

SUPPORTED_FIELDS = frozenset(
    {
        "transaction_at",
        "card",
        "account",
        "owner",
        "institution",
        "account_last4",
        "merchant_raw",
        "amount_aed",
        "spend_aed",
        "currency",
        "amount_original",
        "source_direction",
        "channel",
        "source_type",
        "vendor",
        "category",
        "subcategory",
        "transaction_type",
        "reward_bucket",
        "tags",
        "evidence_policy",
        "evidence_status",
        "review_required",
        "is_refund",
        "is_foreign",
        "is_subscription",
        "property_code",
        "rental_unit",
        "reference",
        "mcc",
        "history_count",
    }
)
SUPPORTED_OPERATORS = frozenset(
    {
        "equals",
        "not_equals",
        "contains",
        "contains_any",
        "not_contains",
        "starts_with",
        "ends_with",
        "regex",
        "in",
        "not_in",
        "numeric_equals",
        "gt",
        "gte",
        "lt",
        "lte",
        "between",
        "date_on",
        "date_before",
        "date_after",
        "date_between",
        "polarity",
        "is_empty",
        "not_empty",
        "is_true",
        "is_false",
        "has_tag",
    }
)
SUPPORTED_ACTIONS = frozenset(
    {"set", "set_if_empty", "add_tag", "add_tags", "remove_tag", "request_evidence", "require_review"}
)
WRITABLE_FIELDS = frozenset(
    {
        "vendor",
        "category",
        "subcategory",
        "transaction_type",
        "reward_bucket",
        "channel",
        "evidence_policy",
        "evidence_status",
        "review_required",
        "is_subscription",
        "owner",
        "property_code",
        "rental_unit",
    }
)


@dataclass(frozen=True, slots=True)
class RuleCondition:
    field: str
    operator: str
    value: Any = None
    second_value: Any = None
    group: int = 1
    negate: bool = False
    case_sensitive: bool = False


@dataclass(frozen=True, slots=True)
class RuleAction:
    action: str
    field: str | None = None
    value: Any = None
    sequence: int = 10


@dataclass(slots=True)
class StaticRule:
    rule_id: str
    name: str
    stage: str
    priority: int
    conditions: list[RuleCondition]
    actions: list[RuleAction]
    enabled: bool = True
    stop_on_match: bool = True
    schema_version: int = SCHEMA_VERSION
    rule_sets: tuple[str, ...] = ("FULL_LEDGER",)

    def canonical(self) -> dict[str, Any]:
        groups: dict[int, list[dict[str, Any]]] = {}
        for condition in self.conditions:
            groups.setdefault(condition.group, []).append(asdict(condition))
        return {
            "schema_version": self.schema_version,
            "rule_id": self.rule_id,
            "name": self.name,
            "stage": self.stage,
            "priority": self.priority,
            "rule_sets": list(self.rule_sets),
            "match": {"any": [{"all": values} for _, values in sorted(groups.items())]},
            "actions": [asdict(action) for action in sorted(self.actions, key=lambda item: item.sequence)],
            "stop_on_match": self.stop_on_match,
        }


@dataclass(frozen=True, slots=True)
class RuleTrace:
    transaction_id: str
    rule_id: str
    rule_name: str
    stage: str
    matched: bool
    actions_applied: tuple[str, ...] = ()


def validate_rule(rule: StaticRule) -> None:
    """Validate the versioned AutoCat-style rule contract before evaluation."""
    errors: list[str] = []
    if rule.schema_version != SCHEMA_VERSION:
        errors.append(f"schema_version must be {SCHEMA_VERSION}")
    if rule.stage not in STAGE_ORDER:
        errors.append(f"unsupported stage {rule.stage!r}")
    if not rule.rule_sets:
        errors.append("at least one rule_set is required")
    for rule_set in rule.rule_sets:
        if not re.fullmatch(r"[A-Z][A-Z0-9_]*", rule_set):
            errors.append(f"invalid rule_set {rule_set!r}")
    if not rule.conditions:
        errors.append("at least one condition is required")
    if not rule.actions:
        errors.append("at least one action is required")

    for index, condition in enumerate(rule.conditions, start=1):
        operator = condition.operator.casefold()
        if condition.field not in SUPPORTED_FIELDS:
            errors.append(f"condition {index} uses unsupported field {condition.field!r}")
        if operator not in SUPPORTED_OPERATORS:
            errors.append(f"condition {index} uses unsupported operator {condition.operator!r}")
        if condition.group < 1:
            errors.append(f"condition {index} group must be at least 1")
        if operator in {"between", "date_between"} and condition.second_value is None:
            errors.append(f"condition {index} {operator} requires second_value")
        if operator == "regex":
            try:
                re.compile(str(condition.value))
            except re.error as error:
                errors.append(f"condition {index} has invalid regex: {error}")

    for index, action in enumerate(rule.actions, start=1):
        action_name = action.action.casefold()
        if action_name not in SUPPORTED_ACTIONS:
            errors.append(f"action {index} uses unsupported action {action.action!r}")
        if action_name in {"set", "set_if_empty"} and action.field not in WRITABLE_FIELDS:
            errors.append(f"action {index} cannot set field {action.field!r}")
        if action_name in {"add_tag", "remove_tag"} and action.value is None:
            errors.append(f"action {index} requires a tag value")
        if action_name == "add_tags" and (
            not isinstance(action.value, (list, tuple, set)) or not action.value
        ):
            errors.append(f"action {index} add_tags requires a non-empty list")

    if errors:
        raise ValueError(f"Invalid rule {rule.rule_id}: " + "; ".join(errors))


def _text(value: Any, case_sensitive: bool) -> str:
    result = "" if value is None else str(value)
    return result if case_sensitive else result.casefold()


def _values(value: Any) -> list[Any]:
    if isinstance(value, (list, tuple, set)):
        return list(value)
    return [value]


def _date(value: Any) -> date:
    if isinstance(value, datetime):
        return value.date()
    if isinstance(value, date):
        return value
    return date.fromisoformat(str(value)[:10])


def condition_matches(transaction: Transaction, condition: RuleCondition) -> bool:
    actual = transaction.value(condition.field)
    op = condition.operator.casefold()
    expected = condition.value
    actual_text = _text(actual, condition.case_sensitive)
    expected_text = _text(expected, condition.case_sensitive)

    if op == "equals":
        result = actual_text == expected_text
    elif op == "not_equals":
        result = actual_text != expected_text
    elif op == "contains":
        result = expected_text in actual_text
    elif op == "contains_any":
        result = any(_text(item, condition.case_sensitive) in actual_text for item in _values(expected))
    elif op == "not_contains":
        result = expected_text not in actual_text
    elif op == "starts_with":
        result = actual_text.startswith(expected_text)
    elif op == "ends_with":
        result = actual_text.endswith(expected_text)
    elif op == "regex":
        flags = 0 if condition.case_sensitive else re.IGNORECASE
        result = re.search(str(expected), "" if actual is None else str(actual), flags) is not None
    elif op == "in":
        result = actual_text in {_text(item, condition.case_sensitive) for item in _values(expected)}
    elif op == "not_in":
        result = actual_text not in {_text(item, condition.case_sensitive) for item in _values(expected)}
    elif op == "numeric_equals":
        result = money(actual) == money(expected)
    elif op in {"gt", "gte", "lt", "lte"}:
        left, right = money(actual), money(expected)
        result = {"gt": left > right, "gte": left >= right, "lt": left < right, "lte": left <= right}[op]
    elif op == "between":
        left = money(actual)
        result = money(expected) <= left <= money(condition.second_value)
    elif op == "date_on":
        result = _date(actual) == _date(expected)
    elif op == "date_before":
        result = _date(actual) < _date(expected)
    elif op == "date_after":
        result = _date(actual) > _date(expected)
    elif op == "date_between":
        result = _date(expected) <= _date(actual) <= _date(condition.second_value)
    elif op == "polarity":
        amount = money(actual)
        result = amount >= 0 if expected_text == "positive" else amount < 0
    elif op == "is_empty":
        result = actual is None or actual == "" or actual == [] or actual == set()
    elif op == "not_empty":
        result = not (actual is None or actual == "" or actual == [] or actual == set())
    elif op == "is_true":
        result = bool(actual)
    elif op == "is_false":
        result = not bool(actual)
    elif op == "has_tag":
        result = expected_text in {_text(item, condition.case_sensitive) for item in transaction.tags}
    else:
        raise ValueError(f"Unsupported rule operator: {condition.operator}")

    return not result if condition.negate else result


def rule_matches(transaction: Transaction, rule: StaticRule) -> bool:
    if not rule.conditions:
        raise ValueError(f"Rule {rule.rule_id} has no conditions")
    groups: dict[int, list[RuleCondition]] = {}
    for condition in rule.conditions:
        groups.setdefault(condition.group, []).append(condition)
    return any(all(condition_matches(transaction, item) for item in conditions) for conditions in groups.values())


def apply_action(transaction: Transaction, action: RuleAction) -> str:
    name = action.action.casefold()
    if name in {"set", "set_if_empty"}:
        if not action.field:
            raise ValueError(f"{name} action requires a field")
        if action.field in set(transaction.metadata.get("locked_fields", [])):
            return f"skipped_locked:{action.field}"
        if name == "set_if_empty" and transaction.value(action.field) not in (None, "", [], set()):
            return f"skipped_populated:{action.field}"
        transaction.set_value(action.field, action.value)
        return f"{name}:{action.field}"
    if name == "add_tag":
        transaction.tags.add(str(action.value))
        return f"add_tag:{action.value}"
    if name == "remove_tag":
        transaction.tags.discard(str(action.value))
        return f"remove_tag:{action.value}"
    if name == "add_tags":
        values = tuple(str(value) for value in _values(action.value))
        transaction.tags.update(values)
        return "add_tags:" + ",".join(values)
    if name == "request_evidence":
        transaction.evidence_policy = str(action.value or "ON_DEMAND")
        transaction.evidence_status = "REQUESTED"
        return "request_evidence"
    if name == "require_review":
        transaction.review_required = True
        return "require_review"
    raise ValueError(f"Unsupported rule action: {action.action}")


class RuleEngine:
    def __init__(self, rules: Iterable[StaticRule]):
        enabled_rules = [rule for rule in rules if rule.enabled]
        for rule in enabled_rules:
            validate_rule(rule)
        self.rules = sorted(
            enabled_rules,
            key=lambda rule: (STAGE_ORDER.get(rule.stage, 999), rule.priority, rule.rule_id),
        )

    def apply(self, transaction: Transaction) -> list[RuleTrace]:
        return self.apply_stages(transaction)

    def apply_stages(
        self,
        transaction: Transaction,
        stages: Iterable[str] | None = None,
    ) -> list[RuleTrace]:
        selected = None if stages is None else {str(stage).upper() for stage in stages}
        traces: list[RuleTrace] = []
        stopped_stages: set[str] = set()
        for rule in self.rules:
            if selected is not None and rule.stage not in selected:
                continue
            if rule.stage in stopped_stages:
                continue
            matched = rule_matches(transaction, rule)
            applied: list[str] = []
            if matched:
                for action in sorted(rule.actions, key=lambda item: item.sequence):
                    applied.append(apply_action(transaction, action))
                if rule.stop_on_match:
                    stopped_stages.add(rule.stage)
            traces.append(
                RuleTrace(
                    transaction_id=transaction.transaction_id,
                    rule_id=rule.rule_id,
                    rule_name=rule.name,
                    stage=rule.stage,
                    matched=matched,
                    actions_applied=tuple(applied),
                )
            )
        return traces


def compile_rules(
    rule_rows: Iterable[dict[str, Any]],
    condition_rows: Iterable[dict[str, Any]],
    action_rows: Iterable[dict[str, Any]],
) -> list[StaticRule]:
    conditions_by_rule: dict[str, list[RuleCondition]] = {}
    for row in condition_rows:
        conditions_by_rule.setdefault(str(row["rule_id"]), []).append(
            RuleCondition(
                field=str(row["field"]),
                operator=str(row["operator"]),
                value=row.get("value"),
                second_value=row.get("second_value"),
                group=int(row.get("group", 1)),
                negate=bool(row.get("negate", False)),
                case_sensitive=bool(row.get("case_sensitive", False)),
            )
        )
    actions_by_rule: dict[str, list[RuleAction]] = {}
    for row in action_rows:
        actions_by_rule.setdefault(str(row["rule_id"]), []).append(
            RuleAction(
                action=str(row["action"]),
                field=row.get("field"),
                value=row.get("value"),
                sequence=int(row.get("sequence", 10)),
            )
        )
    compiled: list[StaticRule] = []
    for row in rule_rows:
        rule_id = str(row["rule_id"])
        raw_rule_sets = row.get("rule_sets") or ("FULL_LEDGER",)
        if isinstance(raw_rule_sets, str):
            raw_rule_sets = (raw_rule_sets,)
        compiled.append(
            StaticRule(
                rule_id=rule_id,
                name=str(row["name"]),
                stage=str(row["stage"]),
                priority=int(row.get("priority", 100)),
                enabled=bool(row.get("enabled", True)),
                stop_on_match=bool(row.get("stop_on_match", True)),
                schema_version=int(row.get("schema_version", SCHEMA_VERSION)),
                rule_sets=tuple(str(value) for value in raw_rule_sets),
                conditions=conditions_by_rule.get(rule_id, []),
                actions=actions_by_rule.get(rule_id, []),
            )
        )
    return compiled

File: packages/n8n-nodes-finance/src/lib/rules.ts
import { JsonObject, assertIsoDate, assertObject, requiredString } from './contracts';
import { RE2JS } from 're2js';

const PROTECTED_FIELDS = new Set([
  'amount', 'amount_aed', 'spend_aed', 'amount_original', 'currency',
  'source_direction', 'transaction_id', 'source_id', 'imported_id',
  'transaction_type', 'transaction_type_locked', 'channel', 'reconciliation_state', 'cashback_finalized', 'locked_fields',
]);
const CONDITION_FIELDS = new Set([
  'transaction_at', 'card', 'account', 'owner', 'institution', 'account_last4',
  'merchant_raw', 'amount', 'amount_aed', 'spend_aed', 'currency', 'amount_original',
  'source_direction', 'channel', 'source_type', 'vendor', 'transaction_type',
  'evidence_policy', 'evidence_status', 'review_required', 'is_refund',
  'is_foreign', 'is_subscription', 'property_code', 'rental_unit', 'reference',
  'mcc', 'history_count', 'tags', 'category', 'subcategory', 'reward_bucket',
]);
const NUMERIC_FIELDS = new Set(['amount', 'amount_aed', 'spend_aed', 'amount_original', 'history_count']);
const BOOLEAN_FIELDS = new Set(['review_required', 'is_refund', 'is_foreign', 'is_subscription']);
const DATE_FIELDS = new Set(['transaction_at']);
const NUMERIC_OPERATORS = new Set(['numeric_equals', 'gt', 'gte', 'lt', 'lte', 'between', 'polarity']);
const BOOLEAN_OPERATORS = new Set(['is_true', 'is_false']);
const TEXT_OPERATORS = new Set(['contains', 'contains_any', 'starts_with', 'ends_with', 'regex']);
const DATE_OPERATORS = new Set(['date_on', 'date_before', 'date_after', 'date_between']);
const N8N_ACTION_FIELDS = new Set([
  'vendor', 'category', 'subcategory', 'reward_bucket', 'evidence_policy', 'evidence_status',
  'review_required', 'is_subscription', 'owner', 'property_code', 'rental_unit',
]);
const STAGE_ORDER = ['TRANSACTION_NORMALIZATION', 'VENDOR_NORMALIZATION', 'CLASSIFICATION', 'TAGGING', 'EVIDENCE', 'CASHBACK'] as const;
const STAGES = new Set<string>(STAGE_ORDER);

export const N8N_RULE_COMPATIBILITY = Object.freeze({
  condition_operators: [
    'equals', 'not_equals', 'contains', 'contains_any', 'not_contains', 'starts_with', 'ends_with',
    'regex', 'in', 'not_in', 'numeric_equals', 'gt', 'gte', 'lt', 'lte', 'between',
    'date_on', 'date_before', 'date_after', 'date_between', 'polarity', 'is_empty',
    'not_empty', 'is_true', 'is_false', 'has_tag',
  ],
  actions: ['set', 'set_if_empty', 'require_review', 'request_evidence', 'add_tag', 'add_tags', 'remove_tag'],
});

export interface NonRepresentableRule {
  schema_version: 1;
  rule_id: string;
  execution_owner: 'N8N_ONLY';
  actual_representable: false;
  stage: string;
  priority: number;
  match: { any: Array<{ all: Array<{ field: string; operator: string; value?: unknown; second_value?: unknown; case_sensitive?: boolean; negate?: boolean }> }> };
  actions: Array<{ action: 'set' | 'set_if_empty' | 'require_review' | 'request_evidence' | 'add_tag' | 'add_tags' | 'remove_tag'; field?: string; value?: unknown; sequence?: number }>;
  stop_on_match: boolean;
}

export function normalizeTransaction(input: unknown): JsonObject {
  assertObject(input, 'transaction');
  const result: JsonObject = { ...input };
  const locked = lockedFields(result);
  if (!locked.has('merchant_raw') && typeof result.merchant_raw === 'string') result.merchant_raw = result.merchant_raw.replace(/\s+/g, ' ').trim();
  if (!locked.has('vendor') && typeof result.vendor === 'string') result.vendor = result.vendor.replace(/\s+/g, ' ').trim();
  if (!locked.has('currency') && typeof result.currency === 'string') result.currency = result.currency.trim().toUpperCase();
  if (!locked.has('tags') && Array.isArray(result.tags)) result.tags = [...new Set(result.tags.filter(value => typeof value === 'string').map(value => value.trim()).filter(Boolean))].sort();
  return result;
}

function lockedFields(row: JsonObject): Set<string> {
  const value = row.locked_fields;
  if (value !== undefined && (!Array.isArray(value) || value.some(field => typeof field !== 'string'))) throw new Error('locked_fields must be an array of field names');
  const locked = new Set<string>((value ?? []) as string[]);
  if (row.transaction_type_locked === true) locked.add('transaction_type');
  return locked;
}

function finiteNumber(value: unknown): number | undefined {
  if (typeof value !== 'number' && typeof value !== 'string') return undefined;
  if (typeof value === 'string' && value.trim() === '') return undefined;
  const result = Number(value);
  return Number.isFinite(result) ? result : undefined;
}

function calendarDate(value: unknown): string | undefined {
  if (typeof value !== 'string' || !/^\d{4}-\d{2}-\d{2}(?:$|[Tt])/.test(value)) return undefined;
  try { return assertIsoDate(value.slice(0, 10), 'rule date'); } catch { return undefined; }
}

function comparable(value: unknown, caseSensitive: boolean): unknown {
  return typeof value === 'string' && !caseSensitive ? value.toLocaleLowerCase() : value;
}

function safeRegexSource(value: unknown): string {
  const source = requiredString(value, 'regex', 512);
  try { RE2JS.compile(source); } catch { throw new Error('regular expression is outside the RE2-compatible linear-time subset'); }
  return source;
}

function ruleFieldValue(row: JsonObject, field: string): unknown {
  if (field === 'is_foreign') return String(row.currency ?? 'AED').toUpperCase() !== 'AED';
  if (field !== 'spend_aed') return row[field];
  const amount = finiteNumber(row.amount_aed);
  if (amount === undefined) return undefined;
  const topic = row.is_refund === true ? 'REFUND' : String(row.transaction_type || 'PURCHASE').toUpperCase();
  const factor = ['REFUND', 'REVERSAL'].includes(topic) ? -1 : ['PURCHASE', 'FEE', 'INTEREST'].includes(topic) ? 1 : ['PAYMENT', 'TRANSFER', 'REWARD_CREDIT', 'INCOME', 'INVESTMENT', 'CREDIT', 'UNRESOLVED_CREDIT'].includes(topic) ? 0 : undefined;
  if (factor === undefined) throw new Error(`unsupported transaction topic: ${topic}`);
  return Math.abs(amount) * factor;
}

function conditionMatches(row: JsonObject, condition: NonRepresentableRule['match']['any'][number]['all'][number]): boolean {
  const actual = comparable(ruleFieldValue(row, condition.field), Boolean(condition.case_sensitive));
  const expected = comparable(condition.value, Boolean(condition.case_sensitive));
  let matched: boolean;
  if (NUMERIC_OPERATORS.has(condition.operator) && finiteNumber(actual) === undefined) return Boolean(condition.negate);
  if (DATE_OPERATORS.has(condition.operator)) {
    const day = calendarDate(actual);
    const target = calendarDate(expected);
    const end = calendarDate(condition.second_value);
    matched = day !== undefined && target !== undefined && (
      condition.operator === 'date_on' ? day === target :
      condition.operator === 'date_before' ? day < target :
      condition.operator === 'date_after' ? day > target : end !== undefined && day >= target && day <= end
    );
    return condition.negate ? !matched : matched;
  }
  switch (condition.operator) {
    case 'equals': matched = actual === expected; break;
    case 'not_equals': matched = actual !== expected; break;
    case 'contains': matched = typeof actual === 'string' && typeof expected === 'string' && actual.includes(expected); break;
    case 'contains_any': matched = Array.isArray(condition.value) && condition.value.some(item => typeof actual === 'string' && actual.includes(String(comparable(item, Boolean(condition.case_sensitive))))); break;
    case 'not_contains': matched = typeof actual === 'string' && typeof expected === 'string' && !actual.includes(expected); break;
    case 'starts_with': matched = typeof actual === 'string' && typeof expected === 'string' && actual.startsWith(expected); break;
    case 'ends_with': matched = typeof actual === 'string' && typeof expected === 'string' && actual.endsWith(expected); break;
    case 'regex': {
      const source = safeRegexSource(condition.value);
      const flags = condition.case_sensitive ? 0 : RE2JS.CASE_INSENSITIVE;
      matched = typeof actual === 'string' && RE2JS.compile(source, flags).test(String(row[condition.field]));
      break;
    }
    case 'in': matched = Array.isArray(condition.value) && condition.value.map(item => comparable(item, Boolean(condition.case_sensitive))).includes(actual); break;
    case 'not_in': matched = Array.isArray(condition.value) && !condition.value.map(item => comparable(item, Boolean(condition.case_sensitive))).includes(actual); break;
    case 'numeric_equals': matched = Number(actual) === Number(expected); break;
    case 'gt': matched = Number(actual) > Number(expected); break;
    case 'gte': matched = Number(actual) >= Number(expected); break;
    case 'lt': matched = Number(actual) < Number(expected); break;
    case 'lte': matched = Number(actual) <= Number(expected); break;
    case 'between': matched = Number(actual) >= Number(expected) && Number(actual) <= Number(condition.second_value); break;
    case 'date_on': matched = String(actual) === String(expected); break;
    case 'date_before': matched = String(actual) < String(expected); break;
    case 'date_after': matched = String(actual) > String(expected); break;
    case 'date_between': matched = String(actual) >= String(expected) && String(actual) <= String(condition.second_value); break;
    case 'polarity': {
      const number = Number(actual);
      matched = expected === 'positive' ? number > 0 : expected === 'negative' ? number < 0 : expected === 'zero' ? number === 0 : false;
      break;
    }
    case 'is_empty': matched = actual === null || actual === undefined || actual === ''; break;
    case 'not_empty': matched = actual !== null && actual !== undefined && actual !== ''; break;
    case 'is_true': matched = actual === true; break;
    case 'is_false': matched = actual === false; break;
    case 'has_tag': {
      const tags = row[condition.field];
      matched = Array.isArray(tags) && tags.map(tag => comparable(tag, Boolean(condition.case_sensitive))).includes(expected);
      break;
    }
    default: throw new Error(`unsupported N8N_ONLY operator: ${condition.operator}`);
  }
  return condition.negate ? !matched : matched;
}

export function validateNonRepresentableRule(value: unknown): NonRepresentableRule {
  assertObject(value, 'rule');
  if (value.schema_version !== 1 || value.execution_owner !== 'N8N_ONLY' || value.actual_representable !== false) {
    throw new Error('rule must explicitly be owned by N8N_ONLY and non-representable in Actual');
  }
  const stage = requiredString(value.stage, 'rule.stage', 64);
  requiredString(value.rule_id, 'rule.rule_id', 128);
  if (typeof value.stop_on_match !== 'boolean') throw new Error('rule.stop_on_match must be boolean');
  if (!STAGES.has(stage)) throw new Error(`stage is not eligible for n8n execution: ${stage}`);
  if (!Number.isInteger(value.priority)) throw new Error('rule.priority must be an integer');
  assertObject(value.match, 'rule.match');
  if (!Array.isArray(value.match.any) || value.match.any.length === 0) throw new Error('rule.match.any must be non-empty');
  for (const group of value.match.any) {
    assertObject(group, 'condition group');
    if (!Array.isArray(group.all) || group.all.length === 0) throw new Error('blank condition groups are invalid');
    for (const condition of group.all) {
      assertObject(condition, 'condition');
      const field = requiredString(condition.field, 'condition.field', 64);
      if (!CONDITION_FIELDS.has(field)) throw new Error(`field is not readable by n8n rules: ${field}`);
      const operator = requiredString(condition.operator, 'condition.operator', 32);
      if (!N8N_RULE_COMPATIBILITY.condition_operators.includes(operator)) throw new Error(`unsupported N8N_ONLY operator: ${operator}`);
      if (NUMERIC_OPERATORS.has(operator) && !NUMERIC_FIELDS.has(field)) throw new Error(`numeric operator is invalid for ${field}`);
      if (BOOLEAN_OPERATORS.has(operator) && !BOOLEAN_FIELDS.has(field)) throw new Error(`boolean operator is invalid for ${field}`);
      if (DATE_OPERATORS.has(operator) && !DATE_FIELDS.has(field)) throw new Error(`date operator is invalid for ${field}`);
      if (TEXT_OPERATORS.has(operator) && (NUMERIC_FIELDS.has(field) || BOOLEAN_FIELDS.has(field))) throw new Error(`text operator is invalid for ${field}`);
      if (operator === 'has_tag' && field !== 'tags') throw new Error('has_tag is valid only for tags');
      if (operator === 'polarity' && !['positive', 'negative', 'zero'].includes(String(condition.value))) throw new Error('polarity requires positive, negative, or zero');
      if (NUMERIC_OPERATORS.has(operator) && operator !== 'polarity') {
        const start = finiteNumber(condition.value);
        const end = finiteNumber(condition.second_value);
        if (start === undefined || (operator === 'between' && (end === undefined || start > end))) throw new Error('numeric condition requires finite values in ascending order');
      }
      if (DATE_OPERATORS.has(operator)) {
        const start = calendarDate(condition.value);
        const end = calendarDate(condition.second_value);
        if (start === undefined || (operator === 'date_between' && (end === undefined || start > end))) throw new Error('date condition requires valid dates in ascending order');
      }
      if (operator === 'regex') safeRegexSource(condition.value);
      if (['contains_any', 'in', 'not_in'].includes(operator) && !Array.isArray(condition.value)) throw new Error(`${operator} requires an array value`);
      if (['between', 'date_between'].includes(operator) && condition.second_value === undefined) throw new Error(`${operator} requires second_value`);
    }
  }
  if (!Array.isArray(value.actions) || value.actions.length === 0) throw new Error('rule.actions must be non-empty');
  for (const action of value.actions) {
    assertObject(action, 'action');
    const kind = requiredString(action.action, 'action.action', 32);
    if (!N8N_RULE_COMPATIBILITY.actions.includes(kind)) throw new Error(`unsupported n8n action: ${kind}`);
    if (action.sequence !== undefined && !Number.isInteger(action.sequence)) throw new Error('action.sequence must be an integer');
    if (['add_tag', 'remove_tag'].includes(kind)) requiredString(action.value, 'action tag', 256);
    if (kind === 'add_tags' && (!Array.isArray(action.value) || action.value.length === 0 || action.value.some(tag => typeof tag !== 'string' || tag.trim() === ''))) throw new Error('add_tags requires non-empty string tags');
    if (kind === 'set' || kind === 'set_if_empty') {
      const field = requiredString(action.field, 'action.field', 64);
      const topicDuringNormalization = field === 'transaction_type' && ['TRANSACTION_NORMALIZATION', 'CLASSIFICATION'].includes(stage);
      if ((!topicDuringNormalization && PROTECTED_FIELDS.has(field)) || (!topicDuringNormalization && !N8N_ACTION_FIELDS.has(field))) throw new Error(`field is not mutable in n8n: ${field}`);
      if (topicDuringNormalization && !['PAYMENT', 'TRANSFER', 'REWARD_CREDIT', 'REFUND', 'REVERSAL', 'FEE', 'PURCHASE', 'INCOME', 'INTEREST', 'INVESTMENT'].includes(String(action.value))) throw new Error('transaction_type action has an invalid canonical topic');
    }
  }
  return value as unknown as NonRepresentableRule;
}

export function applyNonRepresentableRules(input: unknown, values: unknown[]): JsonObject {
  let result = normalizeTransaction(input);
  const locked = lockedFields(result);
  const rules = values.map(validateNonRepresentableRule).sort((left, right) => STAGE_ORDER.indexOf(left.stage as never) - STAGE_ORDER.indexOf(right.stage as never) || left.priority - right.priority || left.rule_id.localeCompare(right.rule_id));
  let stage = '';
  let stopped = false;
  let topicFinalized = result.transaction_type_locked === true;
  for (const rule of rules) {
    if (rule.stage !== stage) {
      if (!topicFinalized && stage === 'TRANSACTION_NORMALIZATION') { result = finalizeTransactionType(result); topicFinalized = true; }
      if (!topicFinalized && rule.stage !== 'TRANSACTION_NORMALIZATION') { result = finalizeTransactionType(result); topicFinalized = true; }
      stage = rule.stage; stopped = false;
    }
    if (stopped) continue;
    const matched = rule.match.any.some(group => group.all.every(condition => conditionMatches(result, condition)));
    if (!matched) continue;
    for (const action of [...rule.actions].sort((a, b) => (a.sequence ?? 10) - (b.sequence ?? 10))) {
      const field = action.action === 'require_review' ? 'review_required' : action.action === 'request_evidence' ? 'evidence_status' : ['add_tag', 'add_tags', 'remove_tag'].includes(action.action) ? 'tags' : action.field!;
      if (locked.has(field) || (field === 'transaction_type' && topicFinalized)) continue;
      if (action.action === 'request_evidence' && locked.has('evidence_policy')) continue;
      if (action.action === 'require_review') result.review_required = true;
      else if (action.action === 'request_evidence') { result.evidence_policy = action.value || 'ON_DEMAND'; result.evidence_status = 'REQUESTED'; }
      else if (['add_tag', 'add_tags', 'remove_tag'].includes(action.action)) {
        const tags = new Set(Array.isArray(result.tags) ? result.tags as string[] : []);
        if (action.action === 'remove_tag') tags.delete(action.value as string);
        else for (const tag of action.action === 'add_tags' ? action.value as string[] : [action.value as string]) tags.add(tag);
        result.tags = [...tags].sort();
      }
      else if (action.action === 'set') result[action.field!] = action.value;
      else if (result[action.field!] === undefined || result[action.field!] === null || result[action.field!] === '') result[action.field!] = action.value;
    }
    if (rule.stop_on_match) stopped = true;
  }
  if (!topicFinalized) result = finalizeTransactionType(result);
  return result;
}

export function finalizeTransactionType(input: JsonObject): JsonObject {
  if (lockedFields(input).has('transaction_type')) return input;
  const value = { ...input };
  const description = String(value.description ?? value.merchant_raw ?? '').toUpperCase();
  const direction = String(value.source_direction ?? value.direction ?? '').toUpperCase();
  let topic = typeof value.transaction_type === 'string' ? value.transaction_type.toUpperCase() : '';
  if (!topic) {
    if (direction === 'CREDIT') {
      if (/(PAYMENT RECEIVED|CREDIT REPAYMENT|CARD REPAYMENT)/.test(description)) topic = 'PAYMENT';
      else if (/CASHBACK|REWARD CREDIT/.test(description)) topic = 'REWARD_CREDIT';
      else topic = 'REFUND';
    } else if (/\bTRANSFER\b/.test(description)) topic = 'TRANSFER';
    else if (/\bFEE\b|^VAT ON/.test(description)) topic = 'FEE';
    else topic = 'PURCHASE';
  }
  if (!['PAYMENT', 'TRANSFER', 'REWARD_CREDIT', 'REFUND', 'REVERSAL', 'FEE', 'PURCHASE', 'INCOME', 'INTEREST', 'INVESTMENT'].includes(topic)) throw new Error(`cannot finalize unknown transaction topic: ${topic}`);
  value.transaction_type = topic;
  value.transaction_type_locked = true;
  return value;
}

export function assertProtectedFieldsUnchanged(before: unknown, after: unknown): void {
  assertObject(before, 'before'); assertObject(after, 'after');
  for (const field of PROTECTED_FIELDS) {
    if (JSON.stringify(before[field]) !== JSON.stringify(after[field])) throw new Error(`protected field was changed: ${field}`);
  }
}
